File: Python/16. 3Sum Closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        res = float('inf')
        nums.sort()
        for i in range(len(nums)-2):
            j = i+1
            k = len(nums)-1

            while j<k:
                #print(i,j,k)
                total = nums[i] + nums[j]+nums[k]
                if abs(total-target) < abs(res-target):
                    res= total
                if total<target:
                    j+=1
                else:
                    k-=1
        return res

File: Java/16. 3Sum Closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        int closestSum = nums[0]+nums[1]+nums[2];
        int n = nums.length;
        Arrays.sort(nums);
        for (int i=0;i<n;i++){
            int j = i+1;
            int k= n-1;

            while (j<k){
                int currSum = nums[i]+nums[j]+nums[k];
                int currDiff = Math.abs(currSum-target);
                int closestDiff = Math.abs(closestSum -target);

                if (currDiff<closestDiff){
                    closestSum = currSum;
                }

                if (currSum<target){
                    j++;
                }
                else if (currSum>target){
                    k--;
                }
                else{
                    return currSum;
                }
            }
        }
        return closestSum;
    }
}